### ISSUE TITLE
[8.0] Wrong week number on Purchase Analysis report, it is advanced by one week.

### ISSUE THREAD

**Issue by dsabrinarg** [odoo/odoo#12179]
Impacted versions: 8.0

Steps to reproduce:
1.- Go to Reporting >> Purchase >> Purchase Analysis
2.- Group by Order Date >> Week

![20160525_092305 jpg 4032x3024](https://cloud.githubusercontent.com/assets/7597456/15543507/cb1bdee4-225a-11e6-9353-9bb25cd85a26.png)

Current behavior:
The report indicates W22 in place of W21, it is advanced by one week.

![purchase analysis odoo](https://cloud.githubusercontent.com/assets/7597456/15543634/3ebf3f80-225b-11e6-94da-a0a9604bdb2e.png)

Expected behavior:
The report should render the value on week 21 given that Odoo uses [ISO week numbering](https://en.wikipedia.org/wiki/ISO_week_date)

**Comment by Yenthe666** [comment]
@mart-e another one regarding weeks and how they're shown (such as https://github.com/odoo/odoo/issues/12068 and https://github.com/odoo/odoo/issues/9629).
Should this one be closed? Quoting from the other topic:

> Duplicate of #9629 (and it depends of the standard you are using, go for ISO week date
!)

**Comment by nim-odoo** [comment]
Hello,

In this case, this is a standard behavior of babel which uses a numbering depending on the locale used:
```python
>>> datetime.datetime.now()
datetime.datetime(2016, 6, 16, 13, 14, 17, 92315)
>>> babel.dates.format_date(datetime.datetime.now(), format="'W'w YYYY", locale='en_US')
u'W25 2016'
>>> babel.dates.format_date(datetime.datetime.now(), format="'W'w YYYY", locale='fr_FR')
u'W24 2016'
```

This is consistent with what was done in the calendar view: https://github.com/odoo/odoo/pull/11522

**Comment by nhomar** [comment]
@nim-odoo 

I made my tests and actually babel has a "problem" try to compute any date and only fr_FR works properly.

Said this: Can you dedicate to solve babel's problem? or wire the behavior in odoo and fix the babel's version?

This is an issue (no matter who is the gulty) and IMHO Odoo's should solve it inside its space or fix the lib.

I can dedicate time but considering we have a Licence I think you should do the job.

**Comment by nim-odoo** [comment]
@odony : since we agreed on this, could you expose you point of view?

**Comment by odony** [comment]
@nhomar could you explain how babel is broken? Can we see your test code? What is the locale you used to test, and what did you get as the result, vs the expected result?
Do you have a reference for a babel bug report mentioning that it is somehow broken?

It's difficult to verify if indeed babel as a "problem" without any test data/scenario.. And babel seems to work fine for our tests, including fr_FR, de_DE, and even en_US... ;-)

Here are some test samples: in runbot 9.0, if you use _en_US_ (United States locale), **June 5th** is shown in **W24**, and this is the [expected result](http://www.timeanddate.com/calendar/custom.html%3Fcountry%3D1%26wno%3D1%26moon%3Don%26hol%3D25). Then if you switch to _de_DE_ (Germany locale), June 5th becomes **W22**, as [expected](http://www.timeanddate.com/calendar/custom.html?year=2016&country=8&wno=1&holm=1&df=1). And so on...

@dsabrinarg what language did you set in your scenario? This looks like English (en_US), so you're getting the week numbers for USA. Could you try again after setting a language that matches your country?

**Comment by dsabrinarg** [comment]
@odony sure, this is the current result for es_MX

![analisis compra odoo](https://cloud.githubusercontent.com/assets/7597456/16276651/eca63912-3874-11e6-8ed8-8b1e69228412.png)

![user settings odoo](https://cloud.githubusercontent.com/assets/7597456/16276654/f2e31ec6-3874-11e6-9c67-d7d5b4e0799d.png)

I was expecting W25 for June 22, 2016

**Comment by nhomar** [comment]
@odony, @moylop260 has a better understanding of the [problem](https://trac-hacks.org/ticket/8175), then he will extend the explanation in order to fit the situation.

Here an Old Babel Issue (not solved yet) and Odoo need a parameter (@moylop will explain in other message the rationale).

**Comment by moylop260** [comment]
This is a very old and polemic matter.

You can see the same discuss from java, excel, python, php...
Example: Babel discuss 6 years ago - https://trac-hacks.org/ticket/8175

The main issue is that the people use the country standard for number, date normal format... but not for the week number.

We can define `Use the week number format of you country`
But the real global situation is'nt it.

`MS Excel` fixed this polemic matter with a extra parameter: [Excel WEEKNUM function](https://support.office.com/en-us/article/WEEKNUM-function-e5c43a03-b4ab-426c-b411-b18c13c75340) in order to avoid spent time in new issues from users and It allow choose you preferred format.

```md
There are two systems used for this function:
 - System 1:
   - The week containing January 1 is the first week of the year, and is numbered week 1. (default)
 - System 2:
   - The week containing the first Thursday of the year is the first week of the year, and is numbered as week 1. This system is the methodology specified in ISO 8601, which is commonly known as the European week numbering system.
```

![screen shot 2016-06-22 at 12 56 34 pm](https://cloud.githubusercontent.com/assets/6644187/16277487/c65405ba-3878-11e6-9557-8ec923f9fcdf.png)

Odoo is right!
Babel is right!
People is wrong but the people use odoo (with babel 😄 )

IMHO we need find a similar fix like as `MS Office`, adding for `res.lang` model an extra parameter for the case of week numbering with the options and process the special case just for `w` parameter.

**Comment by wtaferner** [comment]
@moylop260 
I like your explanation and I think using libs is always a good thing, but giving the administrator the possibility to influence the standard of a lib and the rest of the software is always a good thing if it could be applied consistently :-)

**Comment by odony** [comment]
Thanks for the comments!
Our goal is not to reinvent the wheel to display a calendar in Odoo, especially not by adding more columns/settings. Usability is not about adding more config options, it's about simplifying the UI and doing things right by default.

A summary of the problem might help keep things straight.

---
The problem of week numbers boils down to this: our **calendar view** displays a table of 7-day rows with a week number. In order for this result to make sense, that week number must apply to every day displayed on that row. 

    | M | T | W | T | F | S | S |
 ---|---|---|---|---|---|---|---|
W22 → |    |    | 1 | 2 | 3 | 4 | 5 |
W23 → | 6  | 7  | 8 | 9 | 10 | 11 | 12 |
W24 → | 13  | 14  | 15 | 16 | 17 | 18 | 19 |
W25 → | 20  | 21  | 22 | 23 | 24 | 25 | 26 |
W26 → | 27  | 28 | 29 | 30 |      |      |      |

For **reports/group_by** we simply generate the same week numbers than we use for the calendar view. The fact that we are using different libraries for the read_group (babel) and the calendar view is kind of an implementation detail.

The above table is influenced by 2 parameters:
 1. what day is considered the first day of week  (if this is wrong, days belonging to different week numbers end up mixed on the same row)
 2. what week is considered week 1 of any given year (if this is wrong, all week numbers are offset)

Both of these settings vary from country to country, but there is clearly no need to support **different settings** for a given country. Libraries will adjust to what people use most in the country, and we'll use their default. It is more important to be **consistent**, for example between our calendar view and our read_group results.

BTW it would be interesting to know what "authority" to follow in each country to determine what is the correct value for these 2 parameters. This seems mostly empirical. Wikipedia links to this extensive inventory of the 2 parameters for each locale in the world: http://www.pjh2.de/datetime/weeknumber/wnd.php.

Consequence: if one customer disagrees with what the libraries/Odoo is using for their Locale, there are only 2 solutions:
 1. The libraries are wrong for this country/locale and need to be fixed. We're using mainstream libraries so they should already be mostly correct, but there might be some small errors remaining.
 2. The customer has a special need that is not covered in Odoo. They can switch to a more appropriate Odoo locale, or implement this as a customization if they want. There's a good chance they will have trouble with other software too.

---

Now focusing on Babel, the 2 important parameters above are implemented with the help of 2 settings for each locale:  `first_week_day` (obvious, 0-6 = Mon-Sun), and `min_week_days` (week 1 is the first week of a year with at least `min_week_days` days). 
Babel's implementation of [`get_week_number()`](https://github.com/python-babel/babel/blob/5fe694b57ef730c4154f77a78c5dbe819c6298a5/babel/dates.py#L1456-L1486) is rather simple based on these 2 settings.

If we want to see if Babel is correct for the es_MX locale, we look into [the table](http://www.pjh2.de/datetime/weeknumber/wnd.php) and find that for Mexico:
   - the first day of the week is Sunday → `first_week_day` should be **6**
   - week 1 is the week containing January 1rst  → `min_week_days` should be **1**.

Let's see if Babel is right:

```
>>> babel.core.Locale.parse('es_MX').first_week_day
6
>>> babel.core.Locale.parse('es_MX').min_week_days
1
```
Looks good! So what's the expected week number for June 22, 2016?
```
>>> babel.dates.format_date(date(2016,6,22), format="'W'w YYYY", locale='es_MX')
u'W26 2016'
```

So indeed, Babel is right according to the 2 parameters we found for Mexico, and **now we know why**, and we can go on :smiley: 

---

Coming to this bug report, why would @dsabrinarg or anyone expect June 22, 2016 to be in **Week 25** in Mexico, rather than **26**? Most libraries and online resources I can find agree that in Mexico the week starts on Sunday, and that Week 1 is the week including January 1.

Why does everybody agree on this if the people in Mexico are using a different system?
Which parameter values did you expect instead here:
 - First day of week? Sunday or not?
 - Week 1 is which week? The first week with 1/1, or something else?

If we determine these parameters we can find another Locale that @dsabrinarg should be using instead.

In the original bug report @dsabrinarg mentions that "Odoo is using ISO week numbering" (the default in Europe), but **that is not correct**. There are Odoo users outside Europe, so the system adapts to the week numbering that people expect. **Just for this reason this report seems invalid.**

If you want ISO week numbering in a country that is not using this numbering, please explain why, as this seems to contradict what other customers want?

Finally, if you really want this mixed environment, maybe you could try altering the settings of your `res.lang` entry: use `es_ES` as the "locale code" to use the ISO week numbering, and "es_MX" as the "iso code" to keep using the same translations.

**Comment by moylop260** [comment]
> @odony  Our goal is not to reinvent the wheel to display a calendar in Odoo, especially not by adding more columns/settings. Usability is not about adding more config options, it's about simplifying the UI and doing things right by default.

Just curiosity, Why odoo have the fields `separator format` and `decimal separator` instead of simplifying the UI and doing things right by default based on `locale`?

By example:
![screen shot 2016-07-12 at 5 36 25 pm](https://cloud.githubusercontent.com/assets/6644187/16786006/33c3a4f0-4857-11e6-9fd4-df6ebfffe361.png)

More columns/settings was added:
[separator format and decimal separator](https://github.com/odoo/odoo/blob/ee365bbf98fb1edbc9860c6302bbd8e0d376ef72/openerp/addons/base/res/res_lang.py#L139-L140)

And we have a custom transform:
[intersperse-method](https://github.com/odoo/odoo/blob/ee365bbf98fb1edbc9860c6302bbd8e0d376ef72/openerp/addons/base/res/res_lang.py#L291-L303)

(I don't know, maybe the this answer allow us discover that we had the same issue of WOY in the past but with decimal separator or maybe not)

**Comment by mart-e** [comment]
Actually we are [using locale](https://github.com/odoo/odoo/blob/ee365bbf98fb1edbc9860c6302bbd8e0d376ef72/openerp/addons/base/res/res_lang.py#L88) to get the separator format and decimal separator.

However, we store it on the `res.lang` object to [load it](https://github.com/odoo/odoo/blob/ee365bbf98fb1edbc9860c6302bbd8e0d376ef72/addons/web/controllers/main.py#L574) in the web client but mostly to not force a server administrator to install all locales. When a language is loaded but the locale is not present on the server, we [fallback](https://github.com/odoo/odoo/blob/ee365bbf98fb1edbc9860c6302bbd8e0d376ef72/openerp/addons/base/res/res_lang.py#L57) on the default lang of the server.
This way, I could still load Odoo in Kabyle and configure manually the correct separators without needing to contact the server admin.

**Comment by moylop260** [comment]
@mart-e 
Thanks for reply.
Then, Should we consider the same custom fields and logic for WoY by the same reason?

**Comment by odony** [comment]
@moylop260 There are multiple reasons, some historical, and some technical:
 - The use of Babel in the framework is relatively **recent**. Before we used to do all the formatting of floats and dates ourselves, and it was bad. In particular the built-in Python tools only support a single global Locale, so formatting dates in a Locale-aware manner was not thread-safe and thus not possible. Babel does this nicely and safely.
 - We also need Locale-aware formatting of dates on the **client side** (Javascript), where Babel has no equivalent. We've used other Locale-aware libs like _Date.js_ and _Moment.js_ to imitate it. They support a subset of what Babel can do for dates,  but they do not support floats.

Because we need the same "formatting output" in Javascript and Python, we have to combine:
  - **Dates**: we use Babel in Python, and Date.js/Moment.js in Javascript (auto-configured with their own locale data)
  - **Floats**: we use our own logic for formatting floats in both Python and Javascript (and we provide our own "locale data" in res.lang to help users configure it)

If we had a nice Babel equivalent in Javascript we could also switch Float formatting to a default "Locale-aware" method everywhere. And indeed it would do the right thing and make the UI simpler :-)
